### ISSUE TITLE
[HOLD for payment 2023-10-10] [$500] [Distance] - The system does not display "No result found" when entering an invalid address.

### ISSUE THREAD

**Issue by kbecciv** [Expensify/App#27165]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Go to "Settings."
2. Click on "Profile."
3. Select "Personal details."
4. Choose "Home address."
5. Enter an invalid address in the "address" field and observe that it displays "No result found" in the container.
6. Now click FAB (Plus sign).
7. From the menu, select the "Request Money" option.
8. Opt for the "Distance" option.
9. Click on "Start."
10. Enter the same invalid address in the "address" field as previously done in step 5.
11. Observe that "No result found" is not displayed.

## Expected Result:
Upon entering an invalid address, the system should display "No result found."

## Actual Result:
The system does not display "No result found" when entering an invalid address.

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android / native
- [ ] Android / Chrome
- [ ] iOS / native
- [ ] iOS / Safari
- [x] MacOS / Chrome / Safari
- [ ] MacOS / Desktop

**Version Number:** 1.3.67.2
**Reproducible in staging?:** y
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation

https://github.com/Expensify/App/assets/93399543/dd7d0995-25e8-47a6-9cdb-d2eefa6cfe0f

https://github.com/Expensify/App/assets/93399543/8c870f21-cee2-47bb-ab78-4d2c93877c1c

**Expensify/Expensify Issue URL:**
**Issue reported by:** @usmantariq96
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1694077781824689

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01f72156da7ba00e25</li>
        <li>Upwork Job ID: 1701260699268075520</li>
        <li>Last Price Increase: 2023-09-11</li>
<li>Automatic offers: </li>
<ul>
<li>DylanDylann | Contributor | 26812367</li>
</ul></li>
    </ul>
</details>

**Comment by akinwale** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
When searching for an address in the Distance Request flow, the autocomplete input does not display "No results found" for an address with no results.

### What is the root cause of that problem?
This is due to the fact that `predefinedPlaces` has been set for the AddressSearch component. The predefined places will still be displayed even when there are no results found, until the user stops typing.

### What changes do you think we should make in order to solve the problem?
Update the Google Places Autocomplete module to display the No results found even when the `predefinedPlaces` prop has been set.

### What alternative solutions did you explore? (Optional)
None.

**Comment by needpower** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
A user enters the same address in two places:
(1) Settings -> Profile -> Personal Details -> Home address -> Address line 1
(2)  FAB (Plus sign) -> Request Money -> Distance tab -> Start -> Address

Given the same prompt, field (1) displays "No results found", but field (2) displays a list of selectable addresses.

### What is the root cause of that problem?
Field (1) fires an API request proxied to Google Maps API _with filtering by place's type_. 
Field (2) does the same request _without filtering by place's type_.

Comparison of API requests in two fields, given the same prompt `hhhh`:
Field (1):
![image](https://github.com/Expensify/App/assets/9203524/a86c6ffc-569c-4d51-8e58-a6bd8b1732e8)
![image](https://github.com/Expensify/App/assets/9203524/3c8d00e6-0054-4546-bf55-490bcafe450e)

Field (2):
![image](https://github.com/Expensify/App/assets/9203524/c81557ef-b8c3-4aa4-aff8-aa4d920f7719)
![image](https://github.com/Expensify/App/assets/9203524/ca4b4141-c6c1-40af-9939-da5e1117b37a)

The root cause of the problem is that the address input field in Request Money by Distance doesn't display the same auto-suggested addresses as a home address field in Profile Settings.


### What changes do you think we should make in order to solve the problem?
Add filtering to the API request when users type into the Request Money by Distance address field.


### What alternative solutions did you explore? (Optional)
Depending on the UX requirements, we may do the opposite, i.e., mimic the auto-suggestions of the Request Money by Distance address field in the Profile Settings home address field.

**Comment by burczu** [comment]
I've checked the proposals and done my own investigation and it seems like to obtain exactly the same behaviour in Distance Address field as it is done in Personal Details Address field we need to mix two proposals into one.

If we only follow the @DylanDylann's [proposal](https://github.com/Expensify/App/issues/27165#issuecomment-1714173006) we will get results like below:

![Screenshot 2023-09-12 at 14 27 35](https://github.com/Expensify/App/assets/7771498/7df49b7e-9f3d-48bf-b67e-979ced9dd5ca)

This is because filtering by address is disabled for Distance page, as @needpower pointed out in [his proposal](https://github.com/Expensify/App/issues/27165#issuecomment-1714341141). If we additionally remove `resultTypes=""` property (by default it is set to "address") that is now added to the `AddressSearch` component in `WaypointEditor` then it it will work exactly the same as in Personal Details page.

![Screenshot 2023-09-12 at 14 31 42](https://github.com/Expensify/App/assets/7771498/6d77c27b-a7d5-4577-a25f-467ccc0ba3ca)

That being said, I think we should hire one of them to implement merged solution, and split the payment between them.

🎀 👀 🎀 C+ reviewed

**Comment by trjExpensify** [comment]
@neil-marcellini @hayata-suenaga @tgolen can I get your input on this one? 

My initial take was that it wasn't a bug. If you have recent addresses when you're on the waypoint picker, you should see those instead of "no address found", though maybe there is some merit in being more explicit with "no address found" when your search input doesn't match any of your recent addresses for better feedback than this:

<img width="379" alt="image" src="https://github.com/Expensify/App/assets/16232057/ee707fe9-2ad3-4c86-87fb-d80132e2f3dd">

**Comment by burczu** [comment]
@trjExpensify yeah - I agree that maybe keeping the same behaviour in both places may not be the best option in the first place...

**Comment by needpower** [comment]
@tgolen I think from a UX perspective, users search and select distance addresses much more often than changing a home address. It makes sense to remember previous choices to facilitate the distance search for the case when a user what to use the same route in a money request.
But it doesn't work out the same way for home addresses, because there are no obvious reasons to remember previous home address.

cc @burczu @tjferriss

**Comment by trjExpensify** [comment]
Yeah, I totally agree we don't need recent addresses in the home address search. My only question comes down to whether "no address found" should be displayed if you enter an invalid address into the search bar for feedback that what you're typing is yielding no results. 

I do find this a bit strange:

https://github.com/Expensify/App/assets/16232057/81c2f18f-2a19-4e9b-a44c-a1e497297aec

**Comment by JmillsExpensify** [comment]
Agreed. I find that weird. We should be giving feedback for no results.

**Comment by hayata-suenaga** [comment]
giving an explicit feedback that a result was not found might be a good idea

**Comment by needpower** [comment]
Taking the latest feedback into account, I've updated the proposal and attached an image example: https://github.com/Expensify/App/issues/27165#issuecomment-1714341141

@trjExpensify @burczu @JmillsExpensify @hayata-suenaga @tgolen

**Comment by trjExpensify** [comment]
Copying the image here for ease of reference: 

<img width="456" alt="image" src="https://github.com/Expensify/App/assets/16232057/d1314d3c-1c2f-4b7b-bb1f-41ce6ca52a02">

I think it's better, but a couple of things: 

1. I don't think we intended to use `Recent waypoints` versus `Recent destinations` in our original design for the copy of the title.
2. As it's a title, it shouldn't appear like it's own row with the two green lines above and beneath it.
3. I think the recent destinations shouldn't be shown when you've started entering something in the search input (unless they are a match with your search input, of course). 

Here's our mockup for reference (you can ignore "use current location" that's coming separate): 

![image](https://github.com/Expensify/App/assets/16232057/5a690c5e-0ca8-4a26-840f-cd03ec382b6f)

CC: @shawnborton

**Comment by needpower** [comment]
Didn't know that there was an original mockup for this screen 👍It looks berrer

**Comment by needpower** [comment]
@trjExpensify @burczu 
Updated the original proposal with the new example image. It may be not precise, but good enough for the idea pitch: https://github.com/Expensify/App/issues/27165#issuecomment-1714341141

**Comment by trjExpensify** [comment]
Thanks! I think it's missing this portion yeah: 

>3. I think the recent destinations shouldn't be shown when you've started entering something in the search input (unless they are a match with your search input, of course).

**Comment by needpower** [comment]
@trjExpensify, it seems like a new feature. I think it should be addressed in a new issue, and as an optional prop, because it will change current behavior (maybe current behavior was implemented for a reason). What do you think? Or do you know who can help resolve this point?

**Comment by trjExpensify** [comment]
I don't think you should ever see the recents list and "no results found" combined like in the proposal: 
<img width="512" alt="image" src="https://github.com/Expensify/App/assets/16232057/9097629b-03ff-4ff1-b183-ce56c64c1908">

I think we want to achieve the following: 

1. When the search input is focused and blank, only show the recents list.
2. When you start typing an address the recents list goes away as we show a list of auto-completed addresses based on your search input
3. When the search results return no valid addresses, only show "no results found"

Can we achieve that? CC: @tgolen @shawnborton @JmillsExpensify

**Comment by DylanDylann** [comment]
[Updated proposal with new expected](https://github.com/Expensify/App/issues/27165#issuecomment-1714173006)

I also attach some screenshots after implementing my solution. @trjExpensify Could you help to verify that? Many Thanks

**Comment by needpower** [comment]
@trjExpensify 
> I don't think you should ever see the recent list and "no results found" combined like in the proposal:

That is really what I meant, and the screenshot I showed before may be confusing because it displays "No results found" with no data in the field.

[Updated proposal](https://github.com/Expensify/App/issues/27165#issuecomment-1714341141)

**Comment by trjExpensify** [comment]
Cool, @burczu I think this is ready for a review.

**Comment by burczu** [comment]
@needpower In you proposal you only described what you want to achieve, but I would like to see how you would like to do it so please extend the proposal by showing what part of code will be changed and how.

**Comment by needpower** [comment]
@burczu Updated proposal with the code: https://github.com/Expensify/App/issues/27165#issuecomment-1714341141

**Comment by DylanDylann** [comment]
@burczu What do you think about [my proposal](https://github.com/Expensify/App/issues/27165#issuecomment-1714173006)?

**Comment by trjExpensify** [comment]
@burczu you good with the proposal? Let's keep this moving!

**Comment by burczu** [comment]
@trjExpensify I've just checked updated proposals from @needpower and @DylanDylann and I think we should choose [the one](https://github.com/Expensify/App/issues/27165#issuecomment-1714341141) from @needpower - they are quite similar but he took an active part in the discussion. What do you think?

**Comment by DylanDylann** [comment]
@burczu 2 proposals are same but I propose first and it has big different to the old proposal of @needpower with clear implementation and screenshot. Besides, I also follow any update in this issue.

**Comment by DylanDylann** [comment]
I appreciate @needpower's work. I also spend more time for this issue and give proposal first. @needpower So it is perfect if we can do this work together. It will be fair and we also do that in many issues

**Comment by needpower** [comment]
@trjExpensify @burczu 
I don't mind if you give this task to @DylanDylann, and so the final bounty. We both know the solution,  and took an active part in discussion.

**Comment by hayata-suenaga** [comment]
@needpower @DylanDylann @burczu 

Are we good to go with @DylanDylann 's proposal then?

**Comment by burczu** [comment]
@DylanDylann @needpower I'm also ok with splitting this between you two.

cc @trjExpensify

**Comment by hayata-suenaga** [comment]
I gonna assign @DylanDylann so that they can start working on this

**Comment by hayata-suenaga** [comment]
wait... I wasn't assigned to this issue 🤣 I gonna defer to @youssef-lr's decision. Sorry, I don't know why I thought this is my issue 😅

**Comment by trjExpensify** [comment]
>@DylanDylann @needpower I'm also ok with splitting this between you two.

Sure, how will you guys split up the work though to share the bounty?

**Comment by DylanDylann** [comment]
I can raise PR for this issue

**Comment by needpower** [comment]
@trjExpensify 
The splitting way I see:
- Display header _Recent destinations_ -- changes in `Expensify` repo
- skip predefined places when there is some text in the field. Display _No results found_ -- changes in `react-native-google-places-autocomplete` repo.

And I would pass it to @DylanDylann if you don't mind. Changes are small and I will find another issue to solve.

**Comment by trjExpensify** [comment]
Sure thing, @DylanDylann can take this issue then no worries!

**Comment by DylanDylann** [comment]
@trjExpensify Could you assign me to this issue? Then, I can raise a PR to move forward this issue

**Comment by trjExpensify** [comment]
Yeah, happy to do that because @hayata-suenaga reviewed the proposal as per [here](https://github.com/Expensify/App/issues/27165#issuecomment-1729042792).

**Comment by DylanDylann** [comment]
@design team  @trjExpensify 
Please help to confirm the copy "Recent Destination" for both languages
<img width="998" alt="Screenshot 2023-09-15 at 20 08 29" src="https://github.com/Expensify/App/assets/141406735/4b00753f-3333-4487-82f3-34969bd04de4">

**Comment by shawnborton** [comment]
Not sure if this issue is related to this, but "Recent destination" seems like it has too much vertical spacing. It should appear just as a small label directly on top of the two rows below it.

**Comment by trjExpensify** [comment]
Yeah, it'll be added new here so let's make sure we get it right and we can get a design review on the PR as well.

**Comment by trjExpensify** [comment]
I also agree we shouldn't capitalise the second word. It's `Recent destinations` in the doc.

**Comment by DylanDylann** [comment]
<img width="709" alt="Screenshot 2023-09-22 at 15 07 20" src="https://github.com/Expensify/App/assets/141406735/16392117-eb6b-463b-8eb7-bd222a7d2a43">

@trjExpensify 
> Yeah, it'll be added new here

Do you mean this image? Also, help to confirm the copy for the Spanish

**Comment by DylanDylann** [comment]
<img width="876" alt="Screenshot 2023-09-22 at 15 14 45" src="https://github.com/Expensify/App/assets/141406735/52689087-f450-4f77-af06-a00d60ff28e7">
 
@shawnborton @trjExpensify Is it ok?

**Comment by shawnborton** [comment]
Nice, that feels better to me!

**Comment by DylanDylann** [comment]
@shawnborton  Could help to confirm the copy  `Recent destinations` for the Spanish?

**Comment by shawnborton** [comment]
I think that's something @trjExpensify or @youssef-lr can help out with!

**Comment by youssef-lr** [comment]
Requested translation in Slack, will comment here once I have them @DylanDylann

**Comment by DylanDylann** [comment]
I am working on PR

**Comment by hayata-suenaga** [comment]
@DylanDylann can you link the PR?

**Comment by DylanDylann** [comment]
@burczu These PRs ready to review. I will update video for android and IOS soon
https://github.com/Expensify/react-native-google-places-autocomplete/pull/8
https://github.com/Expensify/App/pull/28124

**Comment by burczu** [comment]
@trjExpensify @youssef-lr We have changes in the `Expensify/react-native-places-autocomplete` here: https://github.com/Expensify/react-native-google-places-autocomplete/pull/8 that I've just approved. Please take a look, because without it we can't move forward with the PR for the App project.

**Comment by trjExpensify** [comment]
Over to you, @youssef-lr!

**Comment by youssef-lr** [comment]
Just merged it @burczu!

**Comment by burczu** [comment]
@youssef-lr Thanks!

cc @DylanDylann

**Comment by mountiny** [comment]
Posting here, there was a bug raised from this new behaviour https://github.com/Expensify/App/issues/28719 Can you please address it here:

If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___
**Issue found when executing PR** https://github.com/Expensify/App/pull/28124
## Action Performed:
Precondition: User has some destinations in recent destination list
1. Go to staging.new.expensify.com
2. Go to + > Request money > Distance
3. Click on Start
4. Click on the address field
5. Enter anything on the address field

## Expected Result:
The entire recent destination list will disappear altogether as soon as there is some value in the input field

## Actual Result:
'Recent destination' in the list disappears first, followed by the list when entering value in the input field

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [x] Android / native
- [x] Android / Chrome
- [x] iOS / native
- [x] iOS / Safari
- [x] MacOS / Chrome / Safari
- [x] MacOS / Desktop

**Version Number:** 1.3.76-1

**Reproducible in staging?:** Yes

**Reproducible in production?:** No

**If this was caught during regression testing, add the test name, ID and link from TestRail:**

**Email or phone of affected tester (no customers):**

**Logs:** https://stackoverflow.com/c/expensify/questions/4856

**Notes/Photos/Videos:** Any additional supporting documentation

https://github.com/Expensify/App/assets/78819774/d04943b4-3e3d-40ce-aa92-1290e0f8c1ab

**Expensify/Expensify Issue URL:**

**Issue reported by:** Applause - Internal Team

**Slack conversation:**

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

**Comment by trjExpensify** [comment]
Oh yeah, hm. What's going on there? It's like showing the recent list again briefly with out the sub-header and then other matches based on the search input. I can repro it: 

https://github.com/Expensify/App/assets/16232057/f0836279-ec7c-46d7-b96c-83a50d66df74

**Comment by mountiny** [comment]
Yeah I think it just default to the recent result as it does not have the backend results yet.

**Comment by trjExpensify** [comment]
👋 Can we get this checklist filled out to proceed with payments please?

**Comment by burczu** [comment]
**BugZero Checklist:** The PR fixing this issue has been merged! The following checklist ([instructions](https://stackoverflow.com/c/expensify/questions/14574)) will need to be completed before the issue can be closed:

* [x]  [@burczu] The PR that introduced the bug has been identified. Link to the PR: I don't this is regression created by any other PR - this was just implemented in other way.
* [x]  [@burczu] The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment: n/a
* [x]  [@burczu] A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion: n/a
* [x]  [@burczu] Determine if we should create a regression test for this bug.
* [ ]  [@burczu] If we decide to create a regression test for the bug, please propose the [regression test](https://github.com/Expensify/App/blob/main/contributingGuides/REGRESSION_TEST_BEST_PRACTICES.md) steps to ensure the same bug will not reach production again.
* [ ]  [@trjExpensify] Link the GH issue for creating/updating the regression test once above steps have been agreed upon:

**Comment by burczu** [comment]
@trjExpensify 

> Again here @burczu, there's no justification noted for this. I'd love to be able to read minds, but unfortunately I can't, so it would be good for you to communicate your reasoning.

I think in this case it's just an UX fix, it was still possible to use distance selectors and the whole request money flow based on distance so the issue didn't break any flow in the App.

**Comment by trjExpensify** [comment]
Got it. I guess I tend to agree that it's a fairly minor improvement to add to the regression test run standalone. I just took a cursory look through the distance requests ones and we're covered for things like auto-complete, placenames and no address found if you try to save, so I think that's sufficient without an explicit one for returning no results.

**Comment by trjExpensify** [comment]
Proceeding on with payments then: 

- $50 to @usmantariq96 for the bug report
- $500 to @DylanDylann for the fix

No urgency bonus to apply as per [here](https://github.com/Expensify/App/issues/27165#issuecomment-1742393798). Once confirmed, I'll proceed to settle up.

**Comment by trjExpensify** [comment]
Settled up with @DylanDylann. @usmantariq96, I've sent you an offer.

**Comment by trjExpensify** [comment]
Settled up with @usmantariq96, closing!

**Comment by DylanDylann** [comment]
> Settled up with @DylanDylann.

sorry @trjExpensify but it seems I wasn't paid out on the job yet. Could you help double check, thank you!

**Comment by trjExpensify** [comment]
Cool yeah, I reached out to yah on Slack as it looked like it failed but you never know with Upwork! I've now settled up with you.
